deepinv/training_utils/trainer.py

```diff
@@ -0,0 +1,722 @@
+                    caption=caption,
+                )
+                log_dict_post_epoch[post_str + " samples"] = images
+                log_dict_post_epoch["step"] = epoch
+                wandb.log(log_dict_post_epoch)
+
+    def save_model(self, epoch, eval_psnr=None):
+        r"""
+        Save the model.
+
+        It saves the model every ``ckp_interval`` epochs.
+
+        :param int epoch: Current epoch.
+        :param None, float eval_psnr: Evaluation PSNR.
+        """
+        if (epoch > 0 and epoch % self.ckp_interval == 0) or epoch + 1 == self.epochs:
+            os.makedirs(str(self.save_path), exist_ok=True)
+            state = {
+                "epoch": epoch,
+                "state_dict": self.model.state_dict(),
+                "loss": self.loss_history,
+                "optimizer": self.optimizer.state_dict(),
+            }
+            if eval_psnr is not None:
+                state["eval_psnr"] = eval_psnr
+            torch.save(
+                state, os.path.join(str(self.save_path), "ckp_{}.pth.tar".format(epoch))
+            )
+
+    def train(
+        self,
+        model: torch.nn.Module,
+        physics: Physics,
+        optimizer: torch.optim.Optimizer,
+        train_dataloader: torch.utils.data.DataLoader,
+        epochs: int = 100,
+        losses: Union[Loss, List[Loss]] = SupLoss(),
+        eval_dataloader=None,
+        scheduler: torch.optim.lr_scheduler.LRScheduler = None,
+    ):
+        r"""
+        Train the model.
+
+        It performs the training process, including the setup, the evaluation, the forward and backward passes,
+        and the visualization.
+
+        :param torch.nn.Module model: Reconstruction network, which can be PnP, unrolled, artifact removal
+            or any other custom reconstruction network.
+        :param deepinv.physics.Physics, list[deepinv.physics.Physics] physics: Forward operator(s) used by the reconstruction network.
+        :param int epochs: Number of training epochs. Default is 100.
+        :param torch.nn.optim.Optimizer optimizer: Torch optimizer for training the network.
+        :param torch.utils.data.DataLoader, list[torch.utils.data.DataLoader] train_dataloader: Train data loader(s) should provide a
+            a signal x or a tuple of (x, y) signal/measurement pairs.
+        :param deepinv.loss.Loss, list[deepinv.loss.Loss] losses: Loss or list of losses used for training the model.
+            :ref:`See the libraries' training losses <loss>`. By default, it uses the supervised mean squared error.
+        :param None, torch.utils.data.DataLoader, list[torch.utils.data.DataLoader] eval_dataloader: Evaluation data loader(s)
+            should provide a signal x or a tuple of (x, y) signal/measurement pairs.
+        :param None, torch.optim.lr_scheduler.LRScheduler scheduler: Torch scheduler for changing the learning rate across iterations.
+        :returns: The trained model.
+        """
+
+        self.epochs = epochs
```

Comment:
Why do you assign class attributes here in `train()` rather than in the class initialisation `Trainer()`? This way, it's less elegant to override class attributes. For example I want to be able to do:
```python
@dataclass
class Trainer2(Trainer):
    model: SpecialModule = None
    model2: SpecialModule = None
```
but now I have to override `train()` as well:
```python
@dataclass
class Trainer2(Trainer):
    def train(*args, model: SpecialModule, model2: SpecialModule, **kwargs):
        self.model = model
        self.model2 = model2
        return super().train(*args, **kwargs)
```

Reply:
Good point - I moved many of the training-instance specific inputs (model, dataloader, optimizer) to the train, thinking that the class creation would only save training hyperparams (plotting, wandb, device, etc).

In this way, we could do something like

```
trainer = Trainer(...) 
model1 = trainer.train(model1, loss1, dataloader1, optimizer1)
model2 = trainer.train(model2, loss2, dataloader2, optimizer2)
```

Although this approach has downsides, like the one you raise, and also that the code gets a bit longer (passing params to two functions instead of 1)

Reply:
I might refactor back to everything in the init